File: aoc/2025/rust/src/bin/day01.rs
use std::{
    error::Error,
    fs::File,
    io::{BufRead, BufReader},
    path::PathBuf,
};

use rust::InputFile;

#[derive(Debug)]
enum Rotation {
    Left(i64),
    Right(i64),
}

impl From<&str> for Rotation {
    fn from(value: &str) -> Self {
        let distance: i64 = value[1..].parse().unwrap();
        match value.chars().next() {
            Some('L') => Rotation::Left(distance),
            Some('R') => Rotation::Right(distance),
            _ => unreachable!(),
        }
    }
}

#[derive(Debug)]
struct Dial {
    position: i64,
    stop_at_zero: usize,
    seen_zero: usize,
}

impl Default for Dial {
    fn default() -> Self {
        Self {
            position: 50,
            stop_at_zero: 0,
            seen_zero: 0,
        }
    }
}

impl Dial {
    fn update(&mut self, rotation: &Rotation) {
        match rotation {
            Rotation::Right(distance) => {
                self.position += distance;
                self.seen_zero += (self.position / 100) as usize;
            }
            Rotation::Left(distance) => {
                if self.position == 0 {
                    self.position -= distance;
                    self.seen_zero += (distance / 100) as usize;
                } else {
                    self.position -= distance;
                    if self.position == 0 {
                        self.seen_zero += 1;
                    } else if self.position < 0 {
                        self.seen_zero += (self.position.abs() / 100) as usize + 1;
                    }
                }
            }
        }
        self.position %= 100;
        self.position += 100;
        self.position %= 100;

        if self.position == 0 {
            self.stop_at_zero += 1;
        }
    }
}

fn solve(path: PathBuf) -> Result<(), Box<dyn Error>> {
    let file = File::open(&path)?;
    let reader = BufReader::new(file);
    let mut rotations = vec![];
    for line in reader.lines() {
        let line = line?;
        let rotation = Rotation::from(line.as_str());
        rotations.push(rotation);
    }
    let mut dial = Dial::default();
    for rotation in rotations {
        dial.update(&rotation);
    }
    println!("Part 1: {}", dial.stop_at_zero);
    println!("Part 2: {}", dial.seen_zero);

    Ok(())
}

fn main() {
    solve(InputFile::example(1).path()).unwrap();
    solve(InputFile::actual(1).path()).unwrap();
}

File: aoc/2025/python/day01.py
from dataclasses import dataclass
from enum import StrEnum
from typing import TextIO
from utils import actual, example


class Direction(StrEnum):
    Left = "L"
    Right = "R"


@dataclass()
class Rotation:
    direction: Direction
    distance: int

    @classmethod
    def from_str(cls, s: str) -> Rotation:
        distance = int(s[1:])
        match s[0]:
            case "L":
                return Rotation(direction=Direction.Left, distance=distance)
            case "R":
                return Rotation(direction=Direction.Right, distance=distance)
            case _:
                raise ValueError(s)


@dataclass()
class Dial:
    position: int
    seen_zero: int
    stop_at_zero: int

    def update(self, rotation: Rotation):
        match rotation.direction:
            case Direction.Left:
                if self.position == 0:
                    self.position -= rotation.distance
                    self.seen_zero += (-self.position) // 100
                else:
                    self.position -= rotation.distance
                    if self.position <= 0:
                        self.seen_zero += (-self.position) // 100 + 1
            case Direction.Right:
                self.position += rotation.distance
                self.seen_zero += self.position // 100

        self.position %= 100
        self.position += 100
        self.position %= 100

        if self.position == 0:
            self.stop_at_zero += 1


def solve(lines: TextIO):
    dial = Dial(50, 0, 0)
    for line in lines:
        rotation = Rotation.from_str(line)
        dial.update(rotation)
    print(dial)


def main():
    solve(example(1))
    solve(actual(1))


if __name__ == "__main__":
    main()
